drgn_tools/lsmod.py

```diff
@@ -0,0 +1,165 @@
+
+    :param prog: drgn program
+    :param parm_val: kernel_param.arg is the paramter value that is converted to big endian
+    :param kp: kernel_param
+    :returns: int after converting from hex
+    """
+    parm_val1 = ""
+    for i in range(0, len(parm_val), 2):
+        parm_val1 = (
+            str(prog.read(kp.arg.value_(), 2).hex())[i : i + 2] + parm_val1
+        )
+    return int(str(parm_val1), 16)
+
+
+def lsmod(prog: Program) -> None:
+    """
+    List loaded modules and their details.
+
+    :param prog: drgn program
+    :returns: None
+    """
+
+    # List all loaded modules
+    print(
+        "MODULE".ljust(25)
+        + "NAME".ljust(20)
+        + "\tSIZE".rjust(15)
+        + "\tREF".rjust(3)
+        + "\tDEPENDENDENT MODULES".rjust(15)
+    )
+    print("-" * 90)
+    for use in list_for_each_entry(
+        "struct module", prog["modules"].address_of_(), "list"
+    ):
+        dep_mod = []
+        for depuse in list_for_each_entry(
+            "struct module_use", use.source_list.address_of_(), "source_list"
+        ):
+            dep_mod.append(depuse.source.name.string_().decode("utf-8"))
+        print(
+            hex(use.mkobj.mod.value_()).ljust(5),
+            "\t",
+            use.name.string_().decode("utf-8").ljust(15),
+            "\t",
+            str(use.core_layout.size.value_()).rjust(10),
+            str(use.refcnt.counter.value_() - 1).rjust(6),
+            "\t",
+            ", ".join(dep_mod),
+        )
+
+    # Module details
+    for use in list_for_each_entry(
+        "struct module", prog["modules"].address_of_(), "list"
+    ):
+        print("\n")
+        print("MODULE NAME:".ljust(15), use.name.string_().decode("utf-8"))
+        print("PARAM COUNT:", str(use.num_kp.value_()))
+        print("ADDRESS    :", hex(use.num_kp.address_of_()))
+        print(
+            "PARAMETER".ljust(25),
+            "ADDRESS".rjust(15),
+            "TYPE".rjust(25),
+            "VALUE".rjust(10),
+        )
+        ker_parm = hex(use.kp.value_())
+        for i in range(0, use.num_kp.value_()):
+            parm_valf = ""
+            parm_val = ""
+            kp = Object(prog, "struct kernel_param", address=int(ker_parm, 16))
+            try:
+                param_type = str(
+                    (
+                        (
+                            (identify_address(kp.ops.get).split(":")[1]).split(
+                                "+"
+                            )[0]
+                        ).split("param_")[1]
+                    )
+                )
+                if "charp" in param_type:
+                    param_type = "charp"
+                    parm_val = str(prog.read(kp.arg.value_(), 2).hex())
+                    if parm_val != "0000":
+                        parm_valf = str(int(parm_val, 16))
+                    elif parm_val == "0000":
+                        parm_valf = "NULL"
+                    else:
+                        parm_valf = "UNKNOWN Str"
```

Comment:
Reading the code, `param_get_charp()` interprets `kp->arg` as a `char **`, and reads that pointer. You would want something like this:
```python
value = Object(prog, "char **", value=kp.arg)
if value[0]:
    string = escape_ascii_string(value[0].string_())
else:
    string = "(null)"
```